fish-rust/build.rs

```diff
@@ -104,12 +114,49 @@ fn detect_features() {
 /// Rust offers fine-grained conditional compilation per-os for the popular operating systems, but
 /// doesn't necessarily include less-popular forks nor does it group them into families more
 /// specific than "windows" vs "unix" so we can conditionally compile code for BSD systems.
-fn detect_bsd() -> miette::Result<bool> {
+fn detect_bsd() -> miette::Result<Option<&'static str>> {
     let uname = std::process::Command::new("uname")
         .output()
         .map_err(|_| miette!("Error executing uname!"))?;
-    Ok(std::str::from_utf8(&uname.stdout)
-        .map(|s| s.to_ascii_lowercase())
-        .map(|s| s.contains("bsd"))
-        .unwrap_or(false))
+    Ok(
+        if std::str::from_utf8(&uname.stdout)
+            .map(|s| s.to_ascii_lowercase())
+            .map(|s| s.contains("bsd"))
+            .unwrap_or(false)
+        {
+            Some("bsd")
+        } else {
+            None
+        },
+    )
+}
+
+fn detect_mb_cur_max() -> miette::Result<Option<&'static str>> {
+    // TODO Respect CXX/CXXFLAGS.
+    let mut child = std::process::Command::new("cc")
+        .args(["-x", "c", "-E", "-pipe", "-"])
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .spawn()
+        .unwrap();
+    {
+        let mut stdin = child.stdin.take().unwrap();
+        stdin
+            .write_all(indoc!(
+                b"#include <stdlib.h>
+                  MB_CUR_MAX"
+            ))
+            .unwrap();
+    }
+    let output = child.wait_with_output().unwrap();
+    match String::from_utf8_lossy(&output.stdout)
+        .lines()
+        .last()
+        .unwrap()
+    {
+        "__mb_cur_max" => Ok(Some("apple_libc")),
+        "(___mb_cur_max())" | "((size_t)___mb_cur_max())" => Ok(Some("bsd_libc")),
+        "(__ctype_get_mb_cur_max ())" => Ok(Some("glibc")),
+        def => panic!("Unsupported definition of MB_CUR_MAX: {}", def),
+    }
```

Comment:
Does this work without also adding these features to `Cargo.toml`? If so, that's really cool. It would make it harder for users to manually specify features that are meant to be auto-detected.

Reply:
I think we'd need to use [cargo:rustc-cfg=KEY[="VALUE"]](https://doc.rust-lang.org/cargo/reference/build-scripts.html#rustc-cfg)
so we set it as cfg value directly, not as feature